### ISSUE TITLE
Clarification about loss function mrcnn_bbox_loss_graph

### ISSUE THREAD

**Issue by jnd77** [matterport/Mask_RCNN#786]
https://github.com/matterport/Mask_RCNN/blob/41e7c596ebb83b05a4154bb0ac7a28e0b9afd017/mrcnn/model.py#L1142-L1146

Based on this implementation, the loss on line 1143 is a matrix of shape (number_boxes, 4).
So when taking the mean, we actually get the sum of all elements divided by (4 * number_boxes).

In the original paper, the loss for each box is defined as the sum over the 4 box components.

So we get a loss 4 times smaller than in the paper.

Has anyone noticed any impact ? Is it worth changing the weights in the config to compensate ?

**Comment by waleedka** [comment]
@jnd77 Which paper are you referring to? I investigated this a bit and here are my findings:

I couldn't find a mention of doing a sum over the 4 box components, neither in the Mask RCNN paper, nor in the Faster RCNN paper. 

In the Faster RCNN paper, they mention that the loss is normalized by the number of sampled anchors (not the number of **positive** sampled anchors). In other words, `loss = sum(box regression losses) / count(sampled anchors)`. Our implementation is slightly different, where it's doing a mean, which translates to `loss = sum(box regression losses) / count(positive sampled anchors)`. Notice the subtle difference in the denominator. 

The authors implementation is reasonable, because it means that an image with very few positive anchors will have smaller loss associated with the anchor refinement. This is an update worth doing. 

On average, positive anchors are 33% of anchors. So, on average, our box refinement loss is 3x *larger* than the one in the paper. 

I checked Facebook's [benchmark implementation](https://github.com/facebookresearch/maskrcnn-benchmark/blob/205fa4e7f1b4da72ab35a210dc717eeeb80802fa/maskrcnn_benchmark/modeling/rpn/loss.py), and it seems to match the paper as well. Although, it introduces a `beta` parameter to smooth L1 that was not mentioned in the paper, which is worth it's own separate discussion.

**Comment by jnd77** [comment]
@waleedka sorry, I was referring to the Mask RCNN paper, and the localization loss defined in equation (2): https://arxiv.org/pdf/1504.08083.pdf

The loss is the sum for each box component: x, y, w and h.

This implementation is doing the average.
